### ISSUE TITLE
Problem on how WriteRecors and WriteRecord works with base classes

### ISSUE THREAD

**Issue by gevorgter** [JoshClose/CsvHelper#1528]
```
public class BaseClass
{
 public string name {get;set;}
}

public class MyClass: BaseClass
{
 public string age {get;set;}
}


BaseClass rec = new MyClass {nane="George,age="20"};
csvWriter.WriteRecord(rec);
```

//it will only output name and not age

**Comment by AltruCoder** [comment]
I don't think I would consider that a bug.  With the statement `BaseClass rec = new MyClass {nane="George,age="20"};` you have converted `MyClass` to `BaseClass`.  You wouldn't have access to `age` unless you converted it back to `MyClass`.

**Comment by gevorgter** [comment]
Well, It goes against how any other serializer works. XmlSerializer or JsonSerializer.
Also at least give me an option to use Type instead of <T>.
Problem is that it defined as WriteRecord<T>(object). Can we get a second method WriteRecord(Type, object)? That way i can create 1 CSV with several types of objects. (Car, SportCart, ...) I understand that each record might/will have different amount of columns.

**Comment by JoshClose** [comment]
I would expect it to output the instantiated class and not the casted class. I'll look into why it's doing that. I don't remember off hand.

**Comment by gevorgter** [comment]
I drilled down into source, the problem is with CsvWriter.GetTypeForREcord< T>
It is using type typeof(T) to get writer. 
And looks like there is a workaround. csvWriter.WriteRecord((object)rec); would work correctly.
Basically CsvWriter.GetTypeForRecords checks if T is object then it does dec.GetType()

**Comment by JoshClose** [comment]
Yes, there are a lot of different types that become an issue. The type, the boxed object version of the type, the type as an interface, the type as a parent type, etc.

I'm trying to figure out a way to get them all to work together.
